zebra-network/src/peer_set/candidate_set.rs

```diff
@@ -317,6 +313,9 @@ where
         };
 
         // SECURITY: rate-limit new candidate connections
+        let current_deadline = self.wait_next_handshake.deadline().max(Instant::now());
+        let mut sleep = sleep_until(current_deadline + constants::MIN_PEER_CONNECTION_INTERVAL);
+        mem::swap(&mut self.wait_next_handshake, &mut sleep);
         sleep.await;
 
         Some(reconnect)
```

Comment:
Would this sometimes return `None`?

> Previously, we were advancing the timer even when we returned `None`.

Reply:
Yep, we only make an outbound connection when we have a candidate address that is ready for a new attempt:

https://github.com/ZcashFoundation/zebra/blob/4aecf0360724e50d51f31ab15e86277e5b8bbd32/zebra-network/src/peer_set/candidate_set.rs#L270-L277

Otherwise, we could make multiple connections to an active peer, or retry the same failing peer very fast, particularly on testnet.